src/main/java/com/cloudbees/jenkins/Cleaner.java

```diff
@@ -29,70 +27,50 @@
  */
 @Extension
 public class Cleaner extends PeriodicWork {
-    private final Set<GitHubRepositoryName> couldHaveBeenRemoved = new HashSet<GitHubRepositoryName>();
+    /**
+     * Queue contains repo names prepared to cleanup.
+     * After configure method on job, trigger calls {@link #onStop(AbstractProject)}
+     * which converts to repo names with help of contributors.
+     *
+     * This queue is thread-safe, so any thread can write or
+     * fetch names to this queue without additional sync
+     */
+    private final Queue<GitHubRepositoryName> сleanQueue = new ConcurrentLinkedQueue<GitHubRepositoryName>();
 
     /**
      * Called when a {@link GitHubPushTrigger} is about to be removed.
      */
-    synchronized void onStop(AbstractProject<?,?> job) {
-        couldHaveBeenRemoved.addAll(GitHubRepositoryNameContributor.parseAssociatedNames(job));
+    /* package */ void onStop(AbstractProject<?, ?> job) {
+        сleanQueue.addAll(GitHubRepositoryNameContributor.parseAssociatedNames(job));
     }
 
     @Override
     public long getRecurrencePeriod() {
-        return TimeUnit2.MINUTES.toMillis(3);
+        return TimeUnit.MINUTES.toMillis(3);
     }
 
+    /**
+     * Each run this work fetches alive repo names (which has trigger for it)
+     * then if names queue is not empty (any job was reconfigured with GH trigger change),
+     * next name passed to {@link WebhookManager} with list of active names to check and unregister old hooks
+     */
     @Override
     protected void doRun() throws Exception {
-        List<GitHubRepositoryName> names;
-        synchronized (this) {// atomically obtain what we need to check
-            names = new ArrayList<GitHubRepositoryName>(couldHaveBeenRemoved);
-            couldHaveBeenRemoved.clear();
-        }
+        URL url = Trigger.all().get(GitHubPushTrigger.DescriptorImpl.class).getHookUrl();
```

Comment:
so when first work finds non empty queue it just skip it. And no one will clear it ever :)